numpy/core/src/multiarray/descriptor.c

```diff
@@ -1165,20 +1175,23 @@ _convert_from_dict(PyObject *obj, int align)
         }
         /* Set the itemsize */
         new->elsize = itemsize;
+    } else {
```

Comment:
This would be easier to follow if this branch came first, as you did for `'aligned'`.